### ISSUE TITLE
Loader order confusion: feature suggestion

### ISSUE THREAD

**Issue by cellog** [fuse-box/fuse-box#237]
One of the confusions I have seen and experienced with every loader is when the order of loaders matters.  It is similar to the confusion when passing in arguments to a function that takes a lot of them:

```js
omg('this', 'function', 'takes', 'a', 'lot', 'of', 'arguments', 'I', 'hope', 'the order is correct')
```

In that case, there are solutions that can remove the need to know the order, for example:
```js
much({ better: 'this', is: 'easier', to: 'maintain' })
```

I wonder if fuse use a declarative approach to loaders if that might be easier to debug?

```js
fsbx.FuseBox.init({
        homeDir: "src/client",
        log: false,
        sourceMap: {
            bundleReference: 'app.js.map',
            outFile: `${devReactFolder}/bundles/app.js.map`
        },
        plugins: {
            JSON: true,
            TypeScriptHelpers: true,
            CSS: true,
            SASS: { outputStyle: 'compressed' },
            CSSResource: { inline: true},
            Custom: { someOpt: 'hi' }
        },
        outFile: `${devReactFolder}/bundles/app.js`,
    }).devServer(">[development.tsx] +process -lazy.ts", opts);
```

This would require a couple of changes

 1. some way to register a plugin
 2. some way to declare plugin order

For the first, it would probably be enough to have the key add Plugin and instantiate
For the second, there can be a default order for built-in plugins, and custom plugins can describe an existing plugin they should precede or leave blank if order is unimportant.

thoughts?

**Comment by nchanged** [comment]
Thanks!
I understand what you mean. And now imagine a case
1) You need to 'fix' this library `jstree/dist/themes/default/style.css` using CSSResourcePlugin
2) You are dealing with raw CSS, which should not go to the CSSResourcePlugin
3) Your primary approach is with SASSPlugin + CSSPlugin which inlines the styles.


```
plugins: [
        [/node_modules.*\.css$/, CSSResourcePlugin({ inline: true }), CSSPlugin() ],
        [/lib.*\.css$/,  CSSPlugin({inline : true}) ],
        [ SassPlugin(), CSSPlugin() ]
    ]
```
Plus, in your case, we need to define all the chains manually, as SASS plugin for example just transpiles, the bundling happens in CSSPlugin. 

But it could become even more complicated. Some people would like to store there stuff into a database for example. 
```
[SaSSPlugin, RethingDb()]
```

In a nutshell, you can pipe anything to anywhere with the current architecture. Objects would work nicely if a project is simple enough.

**Comment by nchanged** [comment]
I am adding `debug` mode, which could be easily activated in the config
```
debug : true
```

Or by having `--debug` in the arguments, which should significantly help with debugging. 
What do you think?

**Comment by cellog** [comment]
I see your point about flexibility, it does seem an array is simpler for that case. 

Also, yes to debug mode. 

I have 2 major long-term interests in fuse box. 

The first is easy debugging of problems caused by fusebox. When a problem happens with webpack, the combination of shoddy documentation, internally complex code and a zillion similar options makes debugging basically "try whatever stack overflow says" the only option. To this end, I'd like to see easy test running and full coverage a first class goal. Have a wallabyjs configuration in the root that anyone can use. Design so that code fits on a single page and does one thing per file, split things up if necessary. Whatever it takes to ensure a Javascript expert new to fusebox can read the code and understand how it works without having to read 10 files. 

The second is graduated complexity and difficulty.  Instead of designing with the most advanced user in mind first, I would like to see configuration start with simple for the most common cases, and complex only when necessary. To that end, I would recommend that plug-ins like sass and less by default pass control to the css plugin, and have a configuration option for experts to avoid this (revising my first suggestion based on new knowledge, in other words). Having a "beginner mode" which is on by default and must be explicitly disabled would be great. It would log everything, give warnings if configuration is Wonky. 

I'm also not a fan of magic, so I am not asking you to do anything that would obscure the way it works :). It's always a delicate balance. 

The original subject bears no relation to what I just said, so I leave it to you to decide whether this should split up or continue as a brainstorm and then generate new issues after things are clearer.

**Comment by cellog** [comment]
Oh and making sure there is only ever 1 way to configure the exact same thing. That kind of simplicity yields much more dynamic complexity.

**Comment by nchanged** [comment]
@cellog thanks so much for your contribution. We should definitely continue discussing it.

I agree with you that arrays might look a bit confusing at first, especially if you need to keep the order in mind. Regarding a "beginner mode" - Object style looks awesome, but what about the complicated cases I mentioned above? How about the export mode, where you need to pipe different CSS files into different plugins. 

A 1 way to configure the exact same thing - yes to it. Having 2 ways Objects for beginners and arrays for experts does not sound good.

Regarding logging, I have started doing it already. Gradually we will cover everything and make it easy to read.

**Comment by cellog** [comment]
Oh I want to be clear: scrap my original idea. Arrays are better. 

What I was imagining was more that SassPlugin would automatically call CSSPlugin (beginner mode) unless explicitly asked not to (expert mode) 

It would also be cool if there was a command to generate a configuration, like create-react-app does (and others). Also a command to upgrade configuration to make it easy to adapt to new design choices

But no to object. Since objects cannot work for expert situations, my original idea is not the right solution.

**Comment by nchanged** [comment]
What about CSSPlugin having it all? 

```
CSSPlugin( { type : "sass | less | stylus", config : {} })
```

**Comment by cellog** [comment]
Now that sounds good. Just have to think about how to keep CSSPlugin from getting too bloated.
Perhaps the CSS bundle idea is something more generic? I am thinking a BundlePlugin which is used to take all the files from the plugins before it in the list and make a single file from it. 

But before I get too excited I should note that I'm getting ahead of myself :) 

I think perhaps instead of "type" the option could be "sass" or "less" and so on which would allow us to pass in options as a secondary step between beginner and expert 

```js
CSSPlugin( { less: true | { options}) 
```

If this kind of thing seems appealing as a general Idea, I am imagining the docs could be organized in those sections as well. Beginner - intermediate - advanced. 

Do let me know if these ideas become too intrusive ;)

**Comment by nchanged** [comment]
@cellog absolutely not, In fact, @devmondo suggested that, and now he is probably reading it and saying to himself "I told him" ;-)

Let's brush the CSSPlugin first (basic functionality) take a look [here](https://github.com/fuse-box/fuse-box/blob/master/src/plugins/CSSplugin.ts#L80), it's getting better. I am removing CSSBundle plugin cuz of its redundancy.

And then we can move on to organising the docs and introducing the N00b mode. ;-)

**Comment by devmondo** [comment]
@nchanged you bet i am 😠

**Comment by cellog** [comment]
Haha OK you two, get a room ;) 

Sounds good. I'm going to keep my distance from the code for a bit, and try to document things as a way of clarifying how it works and also to keep you honest. I'd also like to work on a wallabyjs configuration so test coverage is easier to do real time. At least until you finish your fancy test runner thingy.

I'd like to add one more idea: I think it's wise for there to be logically grouped configuration as we've been talking about here. I don't feel the same way about source. I think having each file do 1 thing is wise both for ease of refactoring and ease of understanding how things click. So extracting smushing several files into one file into a Bundle plugin, but having a "bundle" option that pulls in this file is a good model in my opinion. Keeps the user interface simple and intuitive, keeps developing side simple and intuitive. Win-win. Put another way, this makes deleting code a lot easier, and that will reduce the technical debt down the road.

**Comment by nchanged** [comment]
@cellog I will prepare tests for the CSSPlugin, so we can proceed with the docs ;-)

**Comment by devmondo** [comment]
@cellog just want to say, that we really appreciate your style of detailing your proposals and your humble attitude, we are really being empowered by this kind of behaviour, and we hope that your contribution style set a raw model for others to make FuseBox the ultimate tool :)

**Comment by cellog** [comment]
My pleasure. :)

I came of age (open source-wise) on the wild west of the PHP mailing list, 
so I've learned a few things about not causing nuclear meltdowns, haha :)

**Comment by nchanged** [comment]
Closing this for now. Please feel free to check the latest bleeding edge version `fuse-box@2.0.0-beta.8`
and documentation for it http://fuse-box.org:3333. You are welcome to re-open it elsewise ;-)
